### ISSUE TITLE
lib_ffmpeg: How to set dither method?

### ISSUE THREAD

**Issue by fladd** [pyglet/pyglet#419]
I would like to modify FFmpegSource in such a way that it always resamples 24 bit audio to 16 bit. However, I would also like to tell it to use dithering. I am confused on how to achieve this.

I tried calling
```
avutil.av_opt_set_int(self.audio_convert_ctx, "dither_method", "modified_e_weighted")
```
since those option names are defined in FFmpeg/libswresample/options.c.

However, this returns `-1414549496`, which apparently is FFMPEG telling me that the option does not exist.

I would really appreciate any help here.

**Comment by fladd** [comment]
I am not sure what is going on in the lib_ffmpeg code. It seems that `self.audio_convert_ctx`, which is returned from `swresample.swr_alloc_set_opts` is completely empty. In fact, it doesn't even seem to have a `_fields_` attribute.

**Comment by fladd** [comment]
Likewise, calling

```
avutil.av_opt_find(self.audio_convert_ctx, "in_sample_rate")
```

returns `0`, which means it is not found, even when it has just been set by `swresample.swr_alloc_set_opts` (it is one of the 6 options `swresample.swr_alloc_set_opts` sets).

**Comment by benmoran56** [comment]
Does it look like the ctypes bindings in lib_ffmpeg already exist? If so, then it's possible that FFmpeg was compiled without that particular option. If the ctypes bindings don't exist, then they'll need to be added to get any meaningful results.

**Comment by fladd** [comment]
@benmoran56 Thanks for the reply. Could you elaborate on what you meant with the ctypes binding existing? `avutils` itself is defined in lib_ffmpeg. There is no argtypes defined for the specific function I am calling, but it is my understanding that this is also not necessary, since ctypes will guess the arguments types. Also the fact that the called function actually returns something seems to indicate that the call comes through.
Also, adding the following:
```
avutil.av_opt_set_int.argtypes = [POINTER(SwrContext),
                                                      c_wchar_p,
                                                      c_int,
                                                      c_int]
                    avutil.av_opt_set_int.restype = c_int
```
does not change anything.

Is there anything else I need to do to "bind" these functions correctly?

Any further help to get this working would be really very much appreciated.

**Comment by benmoran56** [comment]
You're on the right track. That's how you write the bindings. Essentially, you need to manually specify all of the argument and return types. It looks like neither the `av_opt_set_int` nor `av_opt_find` functions are defined, so ctypes doesn't know how to handle the call or the return values.

**Comment by fladd** [comment]
Yes, however, it appears that defining arguments and return types (see my post above), does not make any difference. Any other ideas what could still go wrong?

Is it expected that `self.audio_convert_ctx` is completely empty even after `swresample.swr_alloc_set_opts` was called?

**Comment by fladd** [comment]
Thanks again for the suggestions. Unfortunately, I did not manage to make this work, and had to release my **Zipped Audio Player** without dithering support for now (you can find it at https://zipped-album.github.io/ if you are curious).

If you still have some more suggestions on how to achieve dithering when resampling from 24 bit to 16 bit, please let me know, so I can include this in a new version.

Thanks!

**Comment by fladd** [comment]
I think I figured it out. I forgot to wrap the strings in `asbytes`. How silly.

So now this returns `0`, which seems to indicate that there was no error:
```
avutil.av_opt_set(self.audio_convert_ctx,
                  asbytes("dither_method"),
                  asbytes("low_shibata"),
                  0)
```
